python/cudf/cudf/core/dataframe.py

```diff
@@ -1856,36 +1872,53 @@ def rename(self, mapper=None, columns=None, copy=True, inplace=False):
         Notes
         -----
         Difference from pandas:
-          * Support axis='columns' only.
-          * Not supporting: index, level
+            * Not supporting: level
 
         Rename will not overwite column names. If a list with duplicates is
         passed, column names will be postfixed with a number.
         """
-        # Pandas defaults to using columns over mapper
-        if columns:
-            mapper = columns
+        if mapper:
+            if axis == 0:
+                index = mapper
+            elif axis == 1:
+                columns = mapper
 
         out = DataFrame(index=self.index)
-        if isinstance(mapper, Mapping):
+
+        if columns:
             postfix = 1
-            # It is possible for DataFrames with a MultiIndex columns object
-            # to have columns with the same name. The followig use of
-            # _cols.items and ("_1", "_2"... allows the use of
-            # rename in this case
-            for key, col in self._data.items():
-                if key in mapper:
-                    if mapper[key] in out.columns:
-                        out_column = mapper[key] + "_" + str(postfix)
-                        postfix += 1
+            if isinstance(columns, Mapping):
+                # It is possible for DataFrames with a MultiIndex columns
+                # object to have columns with the same name. The followig
+                # use of _cols.items and ("_1", "_2"... allows the use of
+                # rename in this case
+                for key, col in self._data.items():
+                    if key in columns:
+                        if columns[key] in out.columns:
+                            out_column = columns[key] + "_" + str(postfix)
+                            postfix += 1
+                        else:
+                            out_column = columns[key]
+                        out[out_column] = col
                     else:
-                        out_column = mapper[key]
-                    out[out_column] = col
-                else:
-                    out[key] = col
-        elif callable(mapper):
-            for col in self._data.names:
-                out[mapper(col)] = self[col]
+                        out[key] = col
+            elif callable(columns):
+                for col in self._data.names:
+                    out[columns(col)] = self[col]
+        else:
+            out._data = self._data
+
+        if index:
+            out_idx = list(self.index)
```

Comment:
This will be horribly slow from converting a GPU index to a list element by element. We don't want to do this.

Reply:
What is the best way to alter a value in the index then? I've had trouble finding the right way in the docs. In a lot of our tests they are `RangeIndex`. Below you mention we shouldn't need to change the index type, but I'm cannot change the values of numbers in the index to characters (like `pandas` does) unless it's a `StringIndex`.

Reply:
Sorry, didn't fully grok the `replace` link below. Thank you!

Reply:
Changing the dtype is something I didn't think about as we're much more sensitive than Pandas. cc @brandon-b-miller @ayushdg who have spearheaded multiple typecasting behaviors in other parts for discussion / their opinions.

Reply:
I'd say we should yield a GenericIndex if the names the output names are all numeric, but I think if any of the output names are string or object the whole index has to end up as object. The same all-or-nothing behavior applies to floats as well. To make this determination we'd require a scan of the output names however including the ones that weren't replaced. This check up front would probably be a lot of the actual computational work, I think the cast and replace calls are probably fairly fast. Alternatively, we could just always yield a StringIndex in this case and punt on the checking. This is likely the most common case anyways, people like to have sensible, actual names on their columns. 

If you must convert to a list you might be better off using `the_index.to_arrow().to_pylist()`, similar to Series.

Reply:
@brandon-b-miller we'd have different behavior than Pandas then. I.E. say I had an index of `[0, 1, 2, 3, 4]` and did `.replace({1: 'a'})`.

Pandas Result: `[0, 'a', 2, 3, 4]`
cuDF Result: `['0', 'a', '2', '3', '4']`

This seems extremely problematic to me since we don't support actual `object` type columns.

Reply:
Sorry - I guess I meant yielding a string column in most cases, not object. The fact that we don't support object in general means the following casting rule is the usually the best we can do: The output datatype has to be the smallest cuDF compatible type that can reliably hold all the data that needs to be in the output. If a string is going in, we know in the most general case the output array has to be string type, even though in a subset of cases the strings are all safe to cast to numeric. We can care about those kinds of cases as much as we want from my point of view, with the cost being how closely we care about looking at the data to check for safe casting, which comes at the cost of perf. This looks to be the behavior we already enforce on construction:

```
>>> x = cudf.DataFrame({'a':[1,2,3]}, index=[0, 'a', 1])
>>> x.index
StringIndex(['0' 'a' '1'], dtype='object')
>>> 
```

In general I'm not sure we can match the pandas behavior if we want to store the index on the GPU since we have a hard requirement of homogeneous datatypes in columns.

Reply:
Maybe we should error in the above situation, but that's another discussion 😄.

For now, I'm in favor of following the typecasting that we use elsewhere which I believe defers to the source column (i.e. the index having `rename` called against it), where we try to typecast the input to the type of the source column and raise an error otherwise.

Reply:
I'd be in support of that with the one comment that I think we should adjust the error handling in that case to clearly state what the problem is. Just using a numpy datatype to try and cast a string gives 

```
ValueError: could not convert string to float: 'a'
```

Reply:
So, I changed the index handling to be 

```
if index:
    out = DataFrame(index=[index.get(item, item) for item in self.index])
else:
    out = DataFrame(index=self.index)
```
Using a list comprehension seemed acceptable? As you guys outlined above, this means that the index output will be a string index unless all of the values are int. It defaults properly without issue if there aren't strings in the replacement.

To deal with the fact that it would be a bit of a trek, I did this for the tests.

```
@pytest.mark.parametrize("axis", [0, "index"])
def test_dataframe_index_rename(axis):
    pdf = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]})
    gdf = DataFrame.from_pandas(pdf)

    expect = pdf.rename(mapper={1: 5, 2: 6}, axis=axis)
    got = gdf.rename(mapper={1: 5, 2: 6}, axis=axis)

    assert_eq(expect, got)

    # `pandas` can support indexes with mixed values. We automatically convert
    # an index with mixed values to a `StringIndex`. Breaking the comparison.
    expect = pdf.rename(mapper={1: "x", 2: "y"}, axis=axis)
    got = gdf.rename(mapper={1: "x", 2: "y"}, axis=axis)

    assert_eq([str(item) for item in expect.index], list(got.index))

    expect = pdf.rename(index={1: "x", 2: "y"})
    got = gdf.rename(index={1: "x", 2: "y"})

    assert_eq([str(item) for item in expect.index], list(got.index))

    expect = pdf.rename({1: "x", 2: "y"})
    got = gdf.rename({1: "x", 2: "y"})

    assert_eq([str(item) for item in expect.index], list(got.index))
```

So y'all are saying we want to raise an error if someone tries to rename values in a `RangeIndex` with strings? Is that really necessary?

Reply:
Say someone has `RangeIndex(5)` and does `gdf.rename(index={0: 'a'})` then the output index would be: `['a', 1, 2, 3, 4]`. We actually can't support this in cuDF due to heterogenous data types. We'd need to change this to: `['a', '1', '2', '3', '4']` which seems like entirely unintended behavior.

Reply:
@kkraus14 No, the current output index is `['a', '1', '2', '3', '4']`. So in the tests I string convert the index from `pandas` for comparison since the results don't exactly match. I can't see any reason why this implementation would be a problem for the time being, do you?

Reply:
From the perspective of an end user, I wouldn't expect you to convert all of my integers to strings. Code that does `my_series[1]` would throw from a missing index key at this point.

If we support Union types in the future it would give us a clear path to implement these mixed types, but before then we should throw in situations like this.